### PR TITLE
Sharne/enhancevmsecurity

### PR DESCRIPTION
**Pull request by harneshalaka**
### PR creator: Description
Added new chapter on Enhancing VM Security.

### PR creator: Are there any relevant issues/feature requests?

* bsc#...
* jsc#...

### CODE REVIEW

File: xml/vm_security.xml [hunk 1]
```diff
@@ -74,29 +81,28 @@
...
 <procedure>
-  <step><para>To verify whether the system has started with the new kernel, check the response for the command <command>uname -r</command></para>
-  <screen># uname -r
-    6.4.0-150616.coco15sp6-coco</screen>
+  <step><para>To verify whether the system has started with the new kernel, check the response for the command <command>uname -r</command>.</para>
+  <screen>&prompt.sudo; <command>uname -r</command> <replaceable>6.4.0-150616.coco15sp6-coco</replaceable></screen>
   <para>Ensure that the kernel version displayed contains the coco tag.</para></step>
   <step><para>To check the initialization result of the AMD Secure Processor in the kernel log when the kernel is running, run the command:</para>
-  <screen># dmesg | grep -i ccp
+  <screen>&prompt.sudo; <command>dmesg | grep -i ccp</command>
```
Comment (dariavladykina) [inline xml/vm_security.xml:105]: Please remove indentation in this screen, too.

File: xml/vm_security.xml [hunk 2]
```diff
@@ -25,22 +33,21 @@
   <sect1 xml:id="vm-security-enable-confidential-compute-module">
```
Comment (dariavladykina) [inline xml/vm_security.xml:33]: This is about the title of the previous subsection which needs to be sentence-case: _Supported hardware_.
Also, please add "to" to the paragraph below the title: _A system with an AMD EPYC (3rd Gen or newer) is required **to** run AMD SEV-SNP virtual machines_.

**Review by dariavladykina — COMMENTED** [review]
Hi Shalaka, please see some suggestions here. 
I've added some comments where it was possible to add them. They are mainly about sentence-style capitalization of all headings and no indentation in screens.
Thanks!

File: xml/vm_security.xml [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+<info>
+<abstract>
+<para>You can enhance the security of your virtual machines with AMD Secure Encrypted Virtualization-Secure Nested Paging (SEV-SNP). The AMD SEV-SNP feature isolates virtual machines from the host system and other VMs, protecting the data and code. This feature encrypts data and ensures that all changes with the code and data in the VM are detected or tracked. Since this isolates VMs, the other VMs or the host machine are not affected with threats.</para>
+<para>This section explains the steps to enable and use AMD SEV-SNP on your AMD EPYC server with &productname; &productnumber;.</para>
+</abstract>
+<abstract>
+<note>
+<title>Technology preview for &productname; 15 SP6</title>
+<para>
+This feature is shipped as a technology preview in &productname; 15 SP6. The necessary packages are not part of the default installation or repositories. These packages are shipped via confidential compute module.
```
Comment (dariavladykina) [inline xml/vm_security.xml:18]:
```suggestion
This feature is shipped as a technology preview in &productname; 15 SP6. The necessary packages are not part of the default installation or repositories. These packages are shipped via a confidential compute module.
```

File: xml/vm_security.xml [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+<sect1 xml:id="vm-security-hardware-support">
+<title>Supported hardware</title>
+
+<para>
+A system with an AMD EPYC (3rd Gen or newer) is required to run AMD SEV-SNP virtual machines. The BIOS of the AMD machine must provide the necessary options to enable support for confidential computing on the platform.</para>
+</sect1>
+<sect1 xml:id="vm-security-enable-confidential-compute-module">
+<title>Enabling confidential compute module</title>
+
+<para>The necessary packages for AMD SEV-SNP feature are shipped via confidential compute module. You must enable it at system installation time or later via the SUSEConnect command-line tool.</para>
```
Comment (dariavladykina) [inline xml/vm_security.xml:36]:
```suggestion
<para>The necessary packages for AMD SEV-SNP feature are shipped via a confidential compute module. You must enable it at system installation time or later via the SUSEConnect command-line tool.</para>
```

File: xml/vm_security.xml [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+</step>
+<step>
+<para>Click <guilabel>Apply</guilabel> and click <guilabel>Begin Installation</guilabel>.</para><para>This starts the virtual machine and install it according to your settings. The virtual machine boots up once the process is complete and you can verify the AMD SEV-SNP protection.</para>
+</step>
+</procedure>
+</sect1> 
+<sect1 xml:id="vm-verify-amd-sv-snp-vm">
+<title>Verifying the AMD SEV-SNP virtual machine</title>
+<para>
+From the appearance of the virtual machine one can not tell whether it runs in a confidential computing environment or not. But there are several ways to verify that from within the virtual machine.</para>
```
Comment (dariavladykina) [inline xml/vm_security.xml:193]:
```suggestion
From the appearance of the virtual machine, one cannot tell whether it runs in a confidential computing environment. But there are several ways to verify that from within the virtual machine.</para>
```

File: xml/vm_security.xml [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+To protect the virtual machine with AMD SEV-SNP, set the correct firmware by modifying the &lt;os&gt; section as given below:</para>
+<screen>
+&lt;os&gt;
+&lt;type arch="x86_64" machine="pc-q35-8.2"&gt;hvm /type&gt;
+&lt;loader readonly="yes" type="rom"&gt;/usr/share/qemu/ovmf-x86_64-sev.bin /loader&gt;
+&lt;boot dev="hd"/&gt;
+/os&gt; </screen>
+<para>The &lt;loader&gt; line sets the firmware to the SEV version of OVMF.</para>
+</step>
+<step><para>Add a &lt;launchSecurity&gt; section. For AMD SEV-SNP, the section looks like this:</para>
```
Comment (dariavladykina) [inline xml/vm_security.xml:164]:
```suggestion
<step><para>Add a <literal>launchSecurity</literal> section. For AMD SEV-SNP, the section looks like this:</para>
```

File: xml/vm_security.xml [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+</imageobject>
+<imageobject role="html">
+<imagedata fileref="vm_security_create_vm_xml.png" width="75%"/>
+</imageobject>
+</mediaobject>
+</figure>
+</step>
+<step>
+<para>
+To protect the virtual machine with AMD SEV-SNP, set the correct firmware by modifying the &lt;os&gt; section as given below:</para>
```
Comment (dariavladykina) [inline xml/vm_security.xml:155]:
```suggestion
To protect the virtual machine with AMD SEV-SNP, set the correct firmware by modifying the <literal>os</literal> section as given below:</para>
```

File: xml/vm_security.xml [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+<screen>
+&lt;os&gt;
+&lt;type arch="x86_64" machine="pc-q35-8.2"&gt;hvm /type&gt;
+&lt;loader readonly="yes" type="rom"&gt;/usr/share/qemu/ovmf-x86_64-sev.bin /loader&gt;
+&lt;boot dev="hd"/&gt;
+/os&gt; </screen>
+<para>The &lt;loader&gt; line sets the firmware to the SEV version of OVMF.</para>
```
Comment (dariavladykina) [inline xml/vm_security.xml:162]:
```suggestion
<para>The <literal>loader</literal> line sets the firmware to the SEV version of OVMF.</para>
```

File: xml/vm_security.xml [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+</procedure>
+</sect1> 
+<sect1 xml:id="vm-verify-amd-sv-snp-vm">
+<title>Verifying the AMD SEV-SNP virtual machine</title>
+<para>
+From the appearance of the virtual machine one can not tell whether it runs in a confidential computing environment or not. But there are several ways to verify that from within the virtual machine.</para>
+<para>To check the kernel log, run the command:</para>
+<screen>&prompt.sudo; <command>dmesg | grep -i sev-snp</command>
+[ 1.986186] Memory Encryption Features active: AMD SEV SEV-ES SEV-SNP</screen>
+<para>The SEV-SNP feature appears in the kernel log among the active memory encryption features shows that the feature is active for the virtual machine.</para>
```
Comment (dariavladykina) [inline xml/vm_security.xml:197]:
```suggestion
<para>The presence of the SEV-SNP feature in the kernel log, among other active 
memory encryption features, shows that it is active for the virtual machine.</para>
```

File: xml/vm_security.xml [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+<para>There are also cryptographically secure ways for prove the security of the AMD SEV-SNP environment.
```
Comment (dariavladykina) [inline xml/vm_security.xml:198]:
```suggestion
<para>There are also cryptographically secure ways to prove the security of the AMD SEV-SNP environment.
```

File: xml/vm_security.xml [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+<imagedata fileref="vm_security_create_vm_details.png" width="75%"/>
+</imageobject>
+<imageobject role="html">
+<imagedata fileref="vm_security_create_vm_details.png" width="75%"/>
+</imageobject>
+</mediaobject>
+</figure>
+</step>
+<step>
+<para>Click <guilabel>Apply</guilabel> and click <guilabel>Begin Installation</guilabel>.</para><para>This starts the virtual machine and install it according to your settings. The virtual machine boots up once the process is complete and you can verify the AMD SEV-SNP protection.</para>
```
Comment (dariavladykina) [inline xml/vm_security.xml:186]:
```suggestion
<para>Click <guilabel>Apply</guilabel> and click <guilabel>Begin Installation</guilabel>.</para><para>This starts the virtual machine and installs it according to your settings. The virtual machine boots up once the process is complete, and you can verify the AMD SEV-SNP protection.</para>
```

File: xml/vm_security.xml [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+<screen>&lt;launchSecurity type="sev-snp"&gt;
+&lt;policy&gt;0x00030000&lt;/policy&gt;
+&lt;/launchSecurity&gt;</screen></step>
+<step>
+<para>Click <guilabel>Apply</guilabel> and then click the <guilabel>Details</guilabel> tab.</para>
+</step>
+<step>
+<para>Select CPUs in the left-hand list, and set the CPU type to host-model:</para>
+<figure>
+<title><guimenu>Details</guimenu> view of virtual machine configuration</title>
```
Comment (dariavladykina) [inline xml/vm_security.xml:174]:
```suggestion
<title>The <guimenu>Details</guimenu> view of virtual machine configuration</title>
```

File: xml/vm_security.xml [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+</step>
+<step><para>Add a &lt;launchSecurity&gt; section. For AMD SEV-SNP, the section looks like this:</para>
+<screen>&lt;launchSecurity type="sev-snp"&gt;
+&lt;policy&gt;0x00030000&lt;/policy&gt;
+&lt;/launchSecurity&gt;</screen></step>
+<step>
+<para>Click <guilabel>Apply</guilabel> and then click the <guilabel>Details</guilabel> tab.</para>
+</step>
+<step>
+<para>Select CPUs in the left-hand list, and set the CPU type to host-model:</para>
```
Comment (dariavladykina) [inline xml/vm_security.xml:172]:
```suggestion
<para>Select CPUs in the left-hand list and set the CPU type to <literal>host-model</literal>:</para>
```

File: xml/vm_security.xml [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+[ 10.133152] ccp 0000:42:00.1: psp enabled
+[ 10.240817] ccp 0000:42:00.1: SEV firmware update successful
+[ 11.128307] ccp 0000:42:00.1: SEV API:1.55 build:8
+[ 11.135057] ccp 0000:42:00.1: SEV-SNP API:1.55 build:8</screen>
+<para>The message about the SEV-SNP API version indicates the successful initialization of the AMD Secure Processor. Sometimes it happens that these messages do not appear in the kernel log. In this case the BIOS settings or the IOMMU configuration are often the root-cause.</para>
+</step>
+</procedure>
+</sect1>
+<sect1 xml:id="vm-launch-amd-sv-snp-vm">
+<title>Launching an AMD SEV-SNP Virtual Machine</title>
```
Comment (dariavladykina) [inline xml/vm_security.xml:112]:
```suggestion
<title>Launching an AMD SEV-SNP virtual machine</title>
```

**Review by dariavladykina — COMMENTED** [review]
Hi, sending back some suggestions of mine. Thanks!

File: xml/vm_security.xml [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+<step><para>In the fifth step, verify the details and select <guilabel>Customize configuration before install</guilabel>.</para>
+<figure>
+<title>Create Virtual Machine</title>
+<mediaobject>
+<imageobject role="fo">
+<imagedata fileref="vm_security_create_vm.png" width="75%"/>
+</imageobject>
+<imageobject role="html">
+<imagedata fileref="vm_security_create_vm.png" width="75%"/>
+</imageobject>
```
Comment (dariavladykina) [inline xml/vm_security.xml:133]: Please add with alt text for the HTML output like this:
```
<textobject role="description">
   <phrase>SHORT_DESCRIPTION</phrase>
</textobject>
```

File: xml/vm_security.xml [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+<para>In the XML tab, you can edit the XML configuration of the virtual machine used by the &libvirt; back-end.</para>
+<figure>
+<title><guimenu>XML</guimenu> view of virtual machine configuration</title>
+<mediaobject>
+<imageobject role="fo">
+<imagedata fileref="vm_security_create_vm_xml.png" width="75%"/>
+</imageobject>
+<imageobject role="html">
+<imagedata fileref="vm_security_create_vm_xml.png" width="75%"/>
+</imageobject>
```
Comment (dariavladykina) [inline xml/vm_security.xml:149]: Please add with alt text for the HTML output like this:
```
<textobject role="description">
   <phrase>SHORT_DESCRIPTION</phrase>
</textobject>
```

File: xml/vm_security.xml [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+To protect the virtual machine with AMD SEV-SNP, set the correct firmware by modifying the <literal>os</literal> section as given below:</para>
+<figure>
+    <title>Set firmware</title>
+    <mediaobject>
+    <imageobject role="fo">
+    <imagedata fileref="vm_security_xml_os.png" width="75%"/>
+    </imageobject>
+    <imageobject role="html">
+    <imagedata fileref="vm_security_xml_os.png" width="75%"/>
+    </imageobject>
```
Comment (dariavladykina) [inline xml/vm_security.xml:164]: Please add with alt text for the HTML output like this:
```
<textobject role="description">
   <phrase>SHORT_DESCRIPTION</phrase>
</textobject>
```

File: xml/vm_security.xml [hunk 18]
```diff
@@ -0,0 +1,216 @@
...
+<step><para>Add a <literal>launchSecurity</literal> section. For AMD SEV-SNP, the section looks like this:</para>
+    <figure>
+        <title>launchSecurity</title>
+        <mediaobject>
+        <imageobject role="fo">
+        <imagedata fileref="vm_security_xml_launchsecurity.png" width="75%"/>
+        </imageobject>
+        <imageobject role="html">
+        <imagedata fileref="vm_security_xml_launchsecurity.png" width="75%"/>
+        </imageobject>
```
Comment (dariavladykina) [inline xml/vm_security.xml:178]: Please add with alt text for the HTML output like this:
```
<textobject role="description">
   <phrase>SHORT_DESCRIPTION</phrase>
</textobject>
```

File: xml/vm_security.xml [hunk 19]
```diff
@@ -0,0 +1,216 @@
...
+<para>Select CPUs in the left-hand list and set the CPU type to <literal>host-model</literal>:</para>
+<figure>
+<title>The <guimenu>Details</guimenu> view of virtual machine configuration</title>
+<mediaobject>
+<imageobject role="fo">
+<imagedata fileref="vm_security_create_vm_details.png" width="75%"/>
+</imageobject>
+<imageobject role="html">
+<imagedata fileref="vm_security_create_vm_details.png" width="75%"/>
+</imageobject>
```
Comment (dariavladykina) [inline xml/vm_security.xml:195]: Please add with alt text for the HTML output like this:
```
<textobject role="description">
   <phrase>SHORT_DESCRIPTION</phrase>
</textobject>
```

**Review by dariavladykina — COMMENTED** [review]
Hi, please add alternative text to all images to help our SEO. Thanks!

File: xml/vm_security.xml (hunk 15 above)
Comment (harneshalaka) [reply xml/vm_security.xml:135]: Added as given below:
<figure>
<title>Create Virtual Machine</title>
<mediaobject>
<imageobject role="fo">
<imagedata fileref="vm_security_create_vm.png" width="75%"/>
</imageobject>
<textobject role="description"><phrase>Select Customize configuration before install</phrase>
</textobject>
<imageobject role="html">
<imagedata fileref="vm_security_create_vm.png" width="75%"/>
</imageobject>
</mediaobject>
</figure>

File: xml/vm_security.xml (hunk 16 above)
Comment (harneshalaka) [reply xml/vm_security.xml:153]: Added as given below:
<figure>
<title><guimenu>XML</guimenu> view of virtual machine configuration</title>
<mediaobject>
<imageobject role="fo">
<imagedata fileref="vm_security_create_vm_xml.png" width="75%"/>
</imageobject>
<textobject role="description"><phrase>Click XML tab</phrase>
</textobject>
<imageobject role="html">
<imagedata fileref="vm_security_create_vm_xml.png" width="75%"/>
</imageobject>
</mediaobject>
</figure>

File: xml/vm_security.xml (hunk 18 above)
Comment (harneshalaka) [reply xml/vm_security.xml:186]: Added as given below:
<figure>
        <title>launchSecurity</title>
        <mediaobject>
        <imageobject role="fo">
        <imagedata fileref="vm_security_xml_launchsecurity.png" width="75%"/>
        </imageobject>
        <textobject role="description"><phrase>Add launchSecurity</phrase>
    </textobject>
        <imageobject role="html">
        <imagedata fileref="vm_security_xml_launchsecurity.png" width="75%"/>
        </imageobject>
        </mediaobject>
        </figure>

File: xml/vm_security.xml (hunk 17 above)
Comment (harneshalaka) [reply xml/vm_security.xml:170]: Added as given below:
<figure>
    <title>Set firmware</title>
    <mediaobject>
    <imageobject role="fo">
    <imagedata fileref="vm_security_xml_os.png" width="75%"/>
    </imageobject>
    <textobject role="description"><phrase>Set firmware</phrase>
    </textobject>
    <imageobject role="html">
    <imagedata fileref="vm_security_xml_os.png" width="75%"/>
    </imageobject>
    </mediaobject>
    </figure>

File: xml/vm_security.xml (hunk 19 above)
Comment (harneshalaka) [reply xml/vm_security.xml:205]: Added as below:
<figure>
<title>The <guimenu>Details</guimenu> view of virtual machine configuration</title>
<mediaobject>
<imageobject role="fo">
<imagedata fileref="vm_security_create_vm_details.png" width="75%"/>
</imageobject>
<textobject role="description"><phrase>Select CPU model</phrase>
</textobject>
<imageobject role="html">
<imagedata fileref="vm_security_create_vm_details.png" width="75%"/>
</imageobject>
</mediaobject>
</figure>

File: xml/vm_security.xml [hunk 20]
```diff
@@ -0,0 +1,226 @@
...
+</listitem>
+<listitem>
+<para>&libvirt; framework</para>
+</listitem>
+</itemizedlist>
+<procedure>
+<step>
+<para>To install the replacement packages, run the command:</para>
+<screen>
+&prompt.sudo; <command>zypper install --from SLE-Module-Confidential-Computing-15-SP6-Pool --from SLE-Module-Confidential-Computing-15-SP6-Updates qemu libvirt</command></screen>
```
Comment (aginies) [inline xml/vm_security.xml:80]: please add at the end of the commande line: kernel-coco

Comment (harneshalaka) [reply xml/vm_security.xml]: Added:

<command>zypper install --from SLE-Module-Confidential-Computing-15-SP6-Pool --from SLE-Module-Confidential-Computing-15-SP6-Updates qemu libvirt kernel-coco</command>

File: xml/vm_security.xml [hunk 21]
```diff
@@ -0,0 +1,226 @@
...
+[ 11.135057] ccp 0000:42:00.1: SEV-SNP API:1.55 build:8</screen>
+<para>The message about the SEV-SNP API version indicates the successful initialization of the AMD Secure Processor. Sometimes it happens that these messages do not appear in the kernel log. In this case the BIOS settings or the IOMMU configuration are often the root-cause.</para>
+</step>
+</procedure>
+</sect1>
+<sect1 xml:id="vm-launch-amd-sv-snp-vm">
+<title>Launching an AMD SEV-SNP virtual machine</title>
+<para>
+You can run AMD SEV-SNP protected virtual machines using the &libvirt; framework once the confidential computing kernel is booted and the AMD Secure Processor is initialized.</para>
+<para>&libvirt; has several ways of setting up new virtual machines, this document uses a prepared disk image and the virt-manager graphical user interface.</para>
```
Comment (dariavladykina) [inline xml/vm_security.xml:115]:
```suggestion
<para>&libvirt; has several ways of setting up new virtual machines. This document uses a prepared disk image and the virt-manager graphical user interface.</para>
```

**Review by dariavladykina — APPROVED** [review]
LGTM!

**Comment by taroth21** [comment]
@harneshalaka : Thanks, I just fixed two minor issues directly that I found by looking at the PDF. All fine from my side now, please go ahead and squash and merge, then we can backport to maintenance/SLE15SP6 next week. Many thanks!

**Review by taroth21 — APPROVED** [review]
Thank you! Good to go!

File: xml/vm_security.xml [hunk 22]
```diff
@@ -77,7 +77,7 @@ The confidential compute module provides replacement packages supporting AMD SEV
 <step>
 <para>To install the replacement packages, run the command:</para>
 <screen>
-&prompt.sudo; <command>zypper install --from SLE-Module-Confidential-Computing-15-SP6-Pool --from SLE-Module-Confidential-Computing-15-SP6-Updates qemu libvirt</command></screen>
+&prompt.sudo; <command>zypper install --from SLE-Module-Confidential-Computing-15-SP6-Pool --from SLE-Module-Confidential-Computing-15-SP6-Updates qemu libvirt kernel-coco</command></screen>
```
Comment (aginies) [inline xml/vm_security.xml]: And there is still one miussing: qemu-ovmf-x86_64 
Sorry to discover this one so late.

**Comment by taroth21** [comment]
Backported to maintenance/SLE15SP6 with 
- https://github.com/SUSE/doc-sle/commit/9829affc161d2c2650278b9704a0da9b51adb4c8
- https://github.com/SUSE/doc-sle/commit/7380f7db1c466f3cd8e34dc5a69905d1df2d853e
- https://github.com/SUSE/doc-sle/commit/10e67662ba170d4559a21841d73cb145fbb0ac10

**Pull request merged**